README.md
# Comma number
[![Build Status](https://travis-ci.org/elidoran/comma-number.svg?branch=master)](https://travis-ci.org/elidoran/comma-number)
[![npm version](https://badge.fury.io/js/comma-number.svg)](http://badge.fury.io/js/comma-number)
[![Coverage Status](https://coveralls.io/repos/github/elidoran/comma-number/badge.svg?branch=master)](https://coveralls.io/github/elidoran/comma-number?branch=master)


## Install

```sh
$ npm install --save comma-number
```


## Usage

```js
const commaNumber = require('comma-number')

commaNumber(1000) // "1,000"
commaNumber(-1000) // "-1,000"
commaNumber(-1000, '.') // "-1.000"

commaNumber(1000.12) // "1,000.12"
commaNumber(-1000.12) // "-1,000.12"
commaNumber('-1000,12', '.', ',') // "-1.000,12"

// make a new function using custom separator and decimal char:
const format = commaNumber.bindWith('_', '!')
// use it as you would commaNumber().
format(1000)     // "1_000"
format(-1000)    // "-1_000"
format(1000.12)  // "1_000!12"
format(-1000.12) // "-1_000!12"
```


## Version 2 Changes

Revised implementation changes the API a bit:

1. input with a type other than `string` and `number` is returned **as is**, not as `'0'`.
2. supports decimals in the number
3. a string number may use an alternate decimal character, specify it as the third argument
4. added a `bindWith` function to use a currying style to bind options for a reusable format function.

Other changes:

1. Added benchmarking to test implementation performance
2. added code coverage
3. added new badges in this README
4. added more versions to the Travis CI config


## API

### commaNumber(number, [separator=','], [decimalChar='.'])

**Parameters:**

* number : {(Number|String)} Number to format
* separator : {String} Value used to separate numbers
* decimalChar : {String} Value used to separate the decimal value

**Returns:**

* {String} Comma formatted number


### bindWith(separator, decimalChar)

The `commaNumber` function accepts these same parameters as the second and third params. This prevents using currying to bind them and reuse that bound function.

The `bindWith` function accepts the options and returns a function bound with them.

```javascript
// the default commaNumber uses a comma separator and period for decimal char.
var commaNumber = require('comma-number')
  // can build a custom version using bindWith.
  , format = commaNumber.bindWith('_', '!')
  , result1 = commaNumber(1234567.89)
  , result2 = format('1234567.89')

console.log(result1) // outputs:  1,234,567.89
console.log(result2) // outputs:  1_234_567!89
```


## Scripts for Testing, Benchmarking, and Code Coverage

```sh
# run tests via tap
$ npm test

# benchmark current implementation versus previous
npm run benchmark

# get coverage info by default with testing:
npm test
```


## Performance Comparison

The rewrite has a considerable performance increase from the previous version.

I converted the benchmark output from my machine into a [table](docs/benchmark.md).

It compares the performance of version 1.1.0 with 2.0.0. The inputs with decimals can only be processed by the new version so those show as "invalid" for the previous version.


## [MIT License](LICENSE.md)


HISTORY.md
2.1.0 - 2021/04/30
==================

1. tweak the flow a bit so the `separator` and `decimal` value are all in an array which is turned into a string via `join('')`. so, all at once the parts are turned into the final string.
2. added some local scopes to the format.
3. changed how it calculates where to start putting separators.
4. switched a while-loop with an increment in the block to a for-loop with the increment in its loop definition.
5. fixed a bug in the `benchmark/index.js` file when using the `--delta` option.
6. added more (longer) numbers to the tests


2.0.2 - 2021/04/28

1. update dev dependencies
2. add 2021 to copyright
3. reformatted var usage in test/lib/index.js and added new test showing a bug
4. fixed bug when a number input was given for a `decimalChar` other than `'.'`.
5. added node 14 to testing
6. switched from tape to tap for testing and removed istanbul
7. added `.nyc_output` to git ignore
8. updated travis config to use a single VM and use nave to run tests with multiple node versions.
9. add `npm install` to command list in docs/benchmark.md.

2.0.1 - 2019/04/19

1. update dev dependencies
2. add 2018 and 2019 to copyright
3. remove lcov only coverage script
4. add nave dev dependency and use it in test scripts for node 4-12 (evens)
5. use const/let instead of var
6. use more descriptive variable names
7. reformat a bit
8. update benchmark script to report PID and ask before starting so the process can be set to a high priority

2.0.0 - 2017/04/02
==================

1. revised implementation
2. supports decimals
3. more test cases
4. updated travis with modern node versions
5. added badges to README
6. updated README content with change info, new `bindWith()`, benchmark info.
7. added code coverage
8. dropped 'standard' lint
9. added benchmarking
10. added a document showing the benchmark output (docs/benchmark.md)


1.1.0 / 2015-07-16
==================

* Use isFinite instead of Number.isFinite for improved browser compatibility

1.0.0 / 2015-05-12
==================

  * Initial release


test.js
var tap = require('tap')
var commaNumber = require('./index.js')

function testEach(pairs, t, format) {
  t.plan(pairs.length)
  pairs.forEach(function (pair) {
    var input, actual, expected, inputString, description

    input = pair[0]
    actual = format(input)
    expected = pair[1]
    inputString =
      ('object' === typeof input)
      ? JSON.stringify(input)
      : ('string' === typeof input)
        ? '\'' + input + '\''
        : input
    description = inputString + ' => ' + expected

    t.equal(actual, expected, description)
  })
}

tap.test('Formatting', function (t) {

  var testPairs

  testPairs = [
    // Positive numbers
    [0, '0'],
    [1, '1'],
    [12, '12'],
    [123, '123'],
    [1234, '1,234'],
    [12345, '12,345'],
    [123456, '123,456'],
    [1234567, '1,234,567'],
    [12345678, '12,345,678'],
    [123456789, '123,456,789'],
    [1234567890, '1,234,567,890'],
    [12345678901, '12,345,678,901'],
    [123456789012, '123,456,789,012'],
    [1234567890123, '1,234,567,890,123'],
    [Infinity, 'Infinity'],

    // With decimals
    [.1, '0.1'],
    [.12, '0.12'],
    [.123, '0.123'],
    [1.2, '1.2'],
    [1.23, '1.23'],
    [1.234, '1.234'],
    [12.3, '12.3'],
    [12.34, '12.34'],
    [123.4, '123.4'],
    [123.45, '123.45'],
    [1234.5, '1,234.5'],
    [1234.56, '1,234.56'],
    [12345.6, '12,345.6'],
    [12345.67, '12,345.67'],
    [123456.7, '123,456.7'],
    [123456.78, '123,456.78'],
    [123456.789, '123,456.789'],
    [1234567.8, '1,234,567.8'],
    [1234567.89, '1,234,567.89'],
    [1234567.899, '1,234,567.899'],
    [12345678.9, '12,345,678.9'],
    [12345678.99, '12,345,678.99'],
    [12345678.999, '12,345,678.999'],
    [123456789.1, '123,456,789.1'],
    [123456789.12, '123,456,789.12'],
    [123456789.123, '123,456,789.123'],
    [1234567890.1, '1,234,567,890.1'],
    [1234567890.12, '1,234,567,890.12'],
    [1234567890.123, '1,234,567,890.123'],
    [12345678901.2, '12,345,678,901.2'],
    [12345678901.23, '12,345,678,901.23'],
    [12345678901.234, '12,345,678,901.234'],
    [123456789012.3, '123,456,789,012.3'],
    [123456789012.34, '123,456,789,012.34'],
    [123456789012.345, '123,456,789,012.345'],
    [1234567890123.4, '1,234,567,890,123.4'],
    [1234567890123.45, '1,234,567,890,123.45'],
    [1234567890123.456, '1,234,567,890,123.456'],

    // Negative numbers
    [-1, '-1'],
    [-12, '-12'],
    [-123, '-123'],
    [-1234, '-1,234'],
    [-12345, '-12,345'],
    [-123456, '-123,456'],
    [-1234567, '-1,234,567'],
    [-12345678, '-12,345,678'],
    [-123456789, '-123,456,789'],
    [-1234567890, '-1,234,567,890'],
    [-12345678901, '-12,345,678,901'],
    [-123456789012, '-123,456,789,012'],
    [-1234567890123, '-1,234,567,890,123'],
    [-Infinity, '-Infinity'],

    // With decimals
    [-.1, '-0.1'],
    [-.12, '-0.12'],
    [-.123, '-0.123'],
    [-1.2, '-1.2'],
    [-1.23, '-1.23'],
    [-1.234, '-1.234'],
    [-12.3, '-12.3'],
    [-12.34, '-12.34'],
    [-123.4, '-123.4'],
    [-123.45, '-123.45'],
    [-1234.5, '-1,234.5'],
    [-1234.56, '-1,234.56'],
    [-12345.6, '-12,345.6'],
    [-12345.67, '-12,345.67'],
    [-123456.7, '-123,456.7'],
    [-123456.78, '-123,456.78'],
    [-123456.789, '-123,456.789'],
    [-1234567.8, '-1,234,567.8'],
    [-1234567.89, '-1,234,567.89'],
    [-1234567.899, '-1,234,567.899'],
    [-12345678.9, '-12,345,678.9'],
    [-12345678.99, '-12,345,678.99'],
    [-12345678.999, '-12,345,678.999'],
    [-123456789.1, '-123,456,789.1'],
    [-123456789.12, '-123,456,789.12'],
    [-123456789.123, '-123,456,789.123'],
    [-1234567890.1, '-1,234,567,890.1'],
    [-1234567890.12, '-1,234,567,890.12'],
    [-1234567890.123, '-1,234,567,890.123'],
    [-12345678901.2, '-12,345,678,901.2'],
    [-12345678901.23, '-12,345,678,901.23'],
    [-12345678901.234, '-12,345,678,901.234'],
    [-123456789012.3, '-123,456,789,012.3'],
    [-123456789012.34, '-123,456,789,012.34'],
    [-123456789012.345, '-123,456,789,012.345'],
    [-1234567890123.4, '-1,234,567,890,123.4'],
    [-1234567890123.45, '-1,234,567,890,123.45'],
    [-1234567890123.456, '-1,234,567,890,123.456'],

    // Strings
    ['0', '0'],
    ['1', '1'],
    ['12', '12'],
    ['123', '123'],
    ['1234', '1,234'],
    ['12345', '12,345'],
    ['123456', '123,456'],
    ['1234567', '1,234,567'],
    ['12345678', '12,345,678'],
    ['123456789', '123,456,789'],
    ['1234567890', '1,234,567,890'],
    ['12345678901', '12,345,678,901'],
    ['123456789012', '123,456,789,012'],
    ['1234567890123', '1,234,567,890,123'],

    // With decimals
    ['.1', '.1'],
    ['0.1', '0.1'],
    ['.12', '.12'],
    ['0.12', '0.12'],
    ['.123', '.123'],
    ['0.123', '0.123'],
    ['1.2', '1.2'],
    ['1.23', '1.23'],
    ['1.234', '1.234'],
    ['12.3', '12.3'],
    ['12.34', '12.34'],
    ['123.4', '123.4'],
    ['123.45', '123.45'],
    ['1234.5', '1,234.5'],
    ['1234.56', '1,234.56'],
    ['12345.6', '12,345.6'],
    ['12345.67', '12,345.67'],
    ['123456.7', '123,456.7'],
    ['123456.78', '123,456.78'],
    ['123456.789', '123,456.789'],
    ['1234567.8', '1,234,567.8'],
    ['1234567.89', '1,234,567.89'],
    ['1234567.899', '1,234,567.899'],
    ['12345678.9', '12,345,678.9'],
    ['12345678.99', '12,345,678.99'],
    ['12345678.999', '12,345,678.999'],
    ['123456789.1', '123,456,789.1'],
    ['123456789.12', '123,456,789.12'],
    ['123456789.123', '123,456,789.123'],
    ['1234567890.1', '1,234,567,890.1'],
    ['1234567890.12', '1,234,567,890.12'],
    ['1234567890.123', '1,234,567,890.123'],
    ['12345678901.2', '12,345,678,901.2'],
    ['12345678901.23', '12,345,678,901.23'],
    ['12345678901.234', '12,345,678,901.234'],
    ['123456789012.3', '123,456,789,012.3'],
    ['123456789012.34', '123,456,789,012.34'],
    ['123456789012.345', '123,456,789,012.345'],
    ['1234567890123.4', '1,234,567,890,123.4'],
    ['1234567890123.45', '1,234,567,890,123.45'],
    ['1234567890123.456', '1,234,567,890,123.456'],

    // Negative numbers
    ['-1', '-1'],
    ['-12', '-12'],
    ['-123', '-123'],
    ['-1234', '-1,234'],
    ['-12345', '-12,345'],
    ['-123456', '-123,456'],
    ['-1234567', '-1,234,567'],
    ['-12345678', '-12,345,678'],
    ['-123456789', '-123,456,789'],
    ['-1234567890', '-1,234,567,890'],
    ['-12345678901', '-12,345,678,901'],
    ['-123456789012', '-123,456,789,012'],
    ['-1234567890123', '-1,234,567,890,123'],
    [-Infinity, '-Infinity'],

    // With decimals
    ['-.1', '-.1'],
    ['-.12', '-.12'],
    ['-.123', '-.123'],
    ['-1.2', '-1.2'],
    ['-1.23', '-1.23'],
    ['-1.234', '-1.234'],
    ['-12.3', '-12.3'],
    ['-12.34', '-12.34'],
    ['-123.4', '-123.4'],
    ['-123.45', '-123.45'],
    ['-1234.5', '-1,234.5'],
    ['-1234.56', '-1,234.56'],
    ['-12345.6', '-12,345.6'],
    ['-12345.67', '-12,345.67'],
    ['-123456.7', '-123,456.7'],
    ['-123456.78', '-123,456.78'],
    ['-123456.789', '-123,456.789'],
    ['-1234567.8', '-1,234,567.8'],
    ['-1234567.89', '-1,234,567.89'],
    ['-1234567.899', '-1,234,567.899'],
    ['-12345678.9', '-12,345,678.9'],
    ['-12345678.99', '-12,345,678.99'],
    ['-12345678.999', '-12,345,678.999'],
    ['-123456789.1', '-123,456,789.1'],
    ['-123456789.12', '-123,456,789.12'],
    ['-123456789.123', '-123,456,789.123'],
    ['-1234567890.1', '-1,234,567,890.1'],
    ['-1234567890.12', '-1,234,567,890.12'],
    ['-1234567890.123', '-1,234,567,890.123'],
    ['-12345678901.2', '-12,345,678,901.2'],
    ['-12345678901.23', '-12,345,678,901.23'],
    ['-12345678901.234', '-12,345,678,901.234'],
    ['-123456789012.3', '-123,456,789,012.3'],
    ['-123456789012.34', '-123,456,789,012.34'],
    ['-123456789012.345', '-123,456,789,012.345'],
    ['-1234567890123.4', '-1,234,567,890,123.4'],
    ['-1234567890123.45', '-1,234,567,890,123.45'],
    ['-1234567890123.456', '-1,234,567,890,123.456'],

  ]

  testEach(testPairs, t, commaNumber)
  // t.plan(testPairs.length)
  // testPairs.forEach(function (pair) {
  //   var input, actual, expected, inputString, description
  //
  //   input = pair[0]
  //   actual = format(input)
  //   expected = pair[1]
  //   inputString =
  //     ('object' === typeof input)
  //     ? JSON.stringify(input)
  //     : ('string' === typeof input)
  //       ? '\'' + input + '\''
  //       : input
  //   description = inputString + ' => ' + expected
  //
  //   t.equal(actual, expected, description)
  // })
})

tap.test('Invalid input', function (t) {

  var input

  input = [
    [],
    {},
    null,
    undefined,
    'abc',
  ]

  t.plan(6)
  t.equal(commaNumber(input[0]), input[0], '[] => []')
  t.equal(commaNumber(input[1]), input[1], '{} => {}')
  t.equal(commaNumber(input[2]), input[2], 'null => null')
  t.equal(commaNumber(input[3]), input[3], 'undefined => undefined')
  t.equal(commaNumber(input[4]), input[4], '\'abc\' => \'abc\'')

  t.equal(isNaN(commaNumber(NaN)), true, 'NaN => NaN')
})

tap.test('Separator with string inputs', function (t) {
  t.plan(6)
  t.equal(commaNumber('1000', ' '), '1 000', '1000 => 1 000')
  t.equal(commaNumber('1000', '.'), '1.000', '1000 => 1.000')
  t.equal(commaNumber('-1000', '.'), '-1.000', '-1000 => -1.000')
  t.equal(commaNumber('1000.12', ' '), '1 000.12', '1000.12 => 1 000.12')
  t.equal(commaNumber('1000.12', '.'), '1.000.12', '1000.12 => 1.000.12')
  t.equal(commaNumber('-1000.12', '.'), '-1.000.12', '-1000.12 => -1.000.12')
})

tap.test('Separator with number inputs', function (t) {
  t.plan(6)
  t.equal(commaNumber(1000, ' '), '1 000', '1000 => 1 000')
  t.equal(commaNumber(1000, '.'), '1.000', '1000 => 1.000')
  t.equal(commaNumber(-1000, '.'), '-1.000', '-1000 => -1.000')
  t.equal(commaNumber(1000.12, ' '), '1 000.12', '1000.12 => 1 000.12')
  t.equal(commaNumber(1000.12, '.'), '1.000.12', '1000.12 => 1.000.12')
  t.equal(commaNumber(-1000.12, '.'), '-1.000.12', '-1000.12 => -1.000.12')
})

tap.test('Decimal Separator with string inputs', function (t) {
  t.plan(6)
  t.equal(commaNumber('1234.5', undefined, '.'), '1,234.5', '1234.5 => 1,234.5')
  t.equal(commaNumber('1234,5', '.', ','), '1.234,5', '1234,5 => 1.234,5')
  t.equal(commaNumber('1234 5', undefined, ' '), '1,234 5', '1234 5 => 1,234 5')
  t.equal(commaNumber('-1234.5', undefined, '.'), '-1,234.5', '-1234.5 => -1,234.5')
  t.equal(commaNumber('-1234,5', '.', ','), '-1.234,5', '-1234,5 => -1.234,5')
  t.equal(commaNumber('-1234 5', undefined, ' '), '-1,234 5', '-1234 5 => -1,234 5')
})

tap.test('Decimal Separator with number inputs', function (t) {
  t.plan(6)
  t.equal(commaNumber(1234.5, undefined, '.'), '1,234.5', '1234.5 => 1,234.5')
  t.equal(commaNumber(1234.5, '.', ','), '1.234,5', '1234,5 => 1.234,5')
  t.equal(commaNumber(1234.5, undefined, ' '), '1,234 5', '1234 5 => 1,234 5')
  t.equal(commaNumber(-1234.5, undefined, '.'), '-1,234.5', '-1234.5 => -1,234.5')
  t.equal(commaNumber(-1234.5, '.', ','), '-1.234,5', '-1234,5 => -1.234,5')
  t.equal(commaNumber(-1234.5, undefined, ' '), '-1,234 5', '-1234 5 => -1,234 5')
})

tap.test('bindWith() and string inputs', function (t) {

  var boundVersion, testPairs

  boundVersion = commaNumber.bindWith('_', '!')

  testPairs = [
    ['1234!56', '1_234!56'],
    ['1234567!89', '1_234_567!89'],
    ['1234567890!12', '1_234_567_890!12'],
    ['-1234!56', '-1_234!56'],
    ['-1234567!89', '-1_234_567!89'],
    ['-1234567890!12', '-1_234_567_890!12'],
  ]

  testEach(testPairs, t, boundVersion)
})

tap.test('bindWith() and number inputs', function (t) {

  var boundVersion, testPairs

  boundVersion = commaNumber.bindWith('_', '!')

  testPairs = [
    [1234.56, '1_234!56'],
    [1234567.89, '1_234_567!89'],
    [1234567890.12, '1_234_567_890!12'],
    [-1234.56, '-1_234!56'],
    [-1234567.89, '-1_234_567!89'],
    [-1234567890.12, '-1_234_567_890!12'],
  ]

  testEach(testPairs, t, boundVersion)
})


test/index.js
var test = require('tape')
var commaNumber = require('../lib')

function testEach(pairs, t, format) {
  t.plan(pairs.length)
  pairs.forEach(function (pair) {
    var input = pair[0]
      , actual = format(input)
      , expected = pair[1]
      , inputString = typeof input === 'object' ? JSON.stringify(input) :
                      typeof input === 'string' ? '\'' + input + '\''   : input
      , description = inputString + ' => ' + expected

    t.equal(actual, expected, description)
  })
}

test('Formatting', function (t) {

  var testPairs = [
    // Positive numbers
    [0, '0'],
    [1, '1'],
    [12, '12'],
    [123, '123'],
    [1234, '1,234'],
    [12345, '12,345'],
    [123456, '123,456'],
    [1234567, '1,234,567'],
    [1234567890, '1,234,567,890'],
    [Infinity, 'Infinity'],

    // With decimals
    [.1, '0.1'],
    [.12, '0.12'],
    [.123, '0.123'],
    [1.2, '1.2'],
    [1.23, '1.23'],
    [1.234, '1.234'],
    [12.3, '12.3'],
    [12.34, '12.34'],
    [123.4, '123.4'],
    [123.45, '123.45'],
    [1234.5, '1,234.5'],
    [1234.56, '1,234.56'],
    [12345.6, '12,345.6'],
    [12345.67, '12,345.67'],
    [123456.7, '123,456.7'],
    [123456.78, '123,456.78'],
    [123456.789, '123,456.789'],
    [1234567.8, '1,234,567.8'],
    [1234567.89, '1,234,567.89'],
    [1234567.899, '1,234,567.899'],

    // Negative numbers
    [-1, '-1'],
    [-12, '-12'],
    [-123, '-123'],
    [-1234, '-1,234'],
    [-12345, '-12,345'],
    [-123456, '-123,456'],
    [-1234567, '-1,234,567'],
    [-1234567890, '-1,234,567,890'],
    [-Infinity, '-Infinity'],

    // With decimals
    [-.1, '-0.1'],
    [-.12, '-0.12'],
    [-.123, '-0.123'],
    [-1.2, '-1.2'],
    [-1.23, '-1.23'],
    [-1.234, '-1.234'],
    [-12.3, '-12.3'],
    [-12.34, '-12.34'],
    [-123.4, '-123.4'],
    [-123.45, '-123.45'],
    [-1234.5, '-1,234.5'],
    [-1234.56, '-1,234.56'],
    [-12345.6, '-12,345.6'],
    [-12345.67, '-12,345.67'],
    [-123456.7, '-123,456.7'],
    [-123456.78, '-123,456.78'],
    [-123456.789, '-123,456.789'],
    [-1234567.8, '-1,234,567.8'],
    [-1234567.89, '-1,234,567.89'],
    [-1234567.899, '-1,234,567.899'],

    // Strings
    ['0', '0'],
    ['1', '1'],
    ['12', '12'],
    ['123', '123'],
    ['1234', '1,234'],
    ['12345', '12,345'],
    ['123456', '123,456'],
    ['1234567', '1,234,567'],
    ['1234567890', '1,234,567,890'],

    // With decimals
    ['.1', '.1'],
    ['0.1', '0.1'],
    ['.12', '.12'],
    ['0.12', '0.12'],
    ['.123', '.123'],
    ['0.123', '0.123'],
    ['1.2', '1.2'],
    ['1.23', '1.23'],
    ['1.234', '1.234'],
    ['12.3', '12.3'],
    ['12.34', '12.34'],
    ['123.4', '123.4'],
    ['123.45', '123.45'],
    ['1234.5', '1,234.5'],
    ['1234.56', '1,234.56'],
    ['12345.6', '12,345.6'],
    ['12345.67', '12,345.67'],
    ['123456.7', '123,456.7'],
    ['123456.78', '123,456.78'],
    ['123456.789', '123,456.789'],
    ['1234567.8', '1,234,567.8'],
    ['1234567.89', '1,234,567.89'],
    ['1234567.899', '1,234,567.899'],

    // Negative numbers
    ['-1', '-1'],
    ['-12', '-12'],
    ['-123', '-123'],
    ['-1234', '-1,234'],
    ['-12345', '-12,345'],
    ['-123456', '-123,456'],
    ['-1234567', '-1,234,567'],
    ['-1234567890', '-1,234,567,890'],
    [-Infinity, '-Infinity'],

    // With decimals
    ['-.1', '-.1'],
    ['-.12', '-.12'],
    ['-.123', '-.123'],
    ['-1.2', '-1.2'],
    ['-1.23', '-1.23'],
    ['-1.234', '-1.234'],
    ['-12.3', '-12.3'],
    ['-12.34', '-12.34'],
    ['-123.4', '-123.4'],
    ['-123.45', '-123.45'],
    ['-1234.5', '-1,234.5'],
    ['-1234.56', '-1,234.56'],
    ['-12345.6', '-12,345.6'],
    ['-12345.67', '-12,345.67'],
    ['-123456.7', '-123,456.7'],
    ['-123456.78', '-123,456.78'],
    ['-123456.789', '-123,456.789'],
    ['-1234567.8', '-1,234,567.8'],
    ['-1234567.89', '-1,234,567.89'],
    ['-1234567.899', '-1,234,567.899'],


  ]

  testEach(testPairs, t, commaNumber)
  // t.plan(testPairs.length)
  // testPairs.forEach(function (pair) {
  //   var input = pair[0]
  //     , actual = commaNumber(input)
  //     , expected = pair[1]
  //     , inputString = typeof input === 'object' ? JSON.stringify(input) :
  //                     typeof input === 'string' ? '\'' + input + '\''   : input
  //     , description = inputString + ' => ' + expected
  //
  //   t.equal(actual, expected, description)
  // })
})

test('Invalid input', function (t) {

  var input = [
    [],
    {},
    null,
    undefined,
    'abc',
  ]

  t.plan(6)
  t.equal(commaNumber(input[0]), input[0], '[] => []')
  t.equal(commaNumber(input[1]), input[1], '{} => {}')
  t.equal(commaNumber(input[2]), input[2], 'null => null')
  t.equal(commaNumber(input[3]), input[3], 'undefined => undefined')
  t.equal(commaNumber(input[4]), input[4], '\'abc\' => \'abc\'')

  t.equal(isNaN(commaNumber(NaN)), true, 'NaN => NaN')
})

test('Separator', function (t) {
  t.plan(2)
  t.equal(commaNumber(1000, ' '), '1 000', '1000 => 1 000')
  t.equal(commaNumber(1000, '.'), '1.000', '1000 => 1.000')
})

test('Decimal Separator', function (t) {
  t.plan(3)
  t.equal(commaNumber('1234.5', undefined, '.'), '1,234.5', '1234.5 => 1,234.5')
  t.equal(commaNumber('1234,5', '.', ','), '1.234,5', '1234,5 => 1.234,5')
  t.equal(commaNumber('1234 5', undefined, ' '), '1,234 5', '1234 5 => 1,234 5')
})

test('bindWith()', function (t) {

  var boundVersion = commaNumber.bindWith('_', '!')

  var testPairs = [
    ['1234!56', '1_234!56'],
    ['1234567!89', '1_234_567!89'],
    ['1234567890!12', '1_234_567_890!12'],
    ['-1234!56', '-1_234!56'],
    ['-1234567!89', '-1_234_567!89'],
    ['-1234567890!12', '-1_234_567_890!12'],
  ]

  testEach(testPairs, t, boundVersion)
})


index.js
'use strict'

// return a string with the provided number formatted with commas.
// can specify either a Number or a String.
function commaNumber(inputNumber, optionalSeparator, optionalDecimalChar) {

  // default `decimalChar` is a period
  const decimalChar = optionalDecimalChar || '.'

  let stringNumber // we assign this in the switch block and need it later.

  {
    let number // we assign this in the switch block and need it right after.

    switch (typeof inputNumber) {

      case 'string':

        // if there aren't enough digits to need separators then return it
        // NOTE: some numbers which are too small will get passed this
        //       when they have decimal values which make them too long here.
        //       but, the number value check after this switch will catch it.
        if (inputNumber.length < (inputNumber[0] === '-' ? 5 : 4)) {
          return inputNumber
        }

        // remember it as a string in `stringNumber` and convert to a Number
        stringNumber = inputNumber

        // if they're not using the Node standard decimal char then replace it
        // before converting.
        number = Number(
          (decimalChar !== '.') ? stringNumber.replace(decimalChar, '.') : stringNumber
        )
        break

      // convert to a string.
      // NOTE: don't check if the number is too small before converting
      //       because we'll need to return `stringNumber` anyway.
      case 'number':
        stringNumber = String(inputNumber)
        number       = inputNumber
        // create the string version with the decimalChar they specified.
        // this matches what the above case 'string' produces,
        // and, fixes the bug *not* doing this caused.
        if ('.' !== decimalChar && !Number.isInteger(inputNumber)) {
          stringNumber = stringNumber.replace('.', decimalChar)
        }
        break

      // return invalid type as-is
      default: return inputNumber
    }

    // when it doesn't need a separator or isn't a number then return it
    if ((-1000 < number && number < 1000) || isNaN(number) || !isFinite(number)) {
      return stringNumber
    }
  }

  {
    // strip off decimal value to add back in later
    const decimalIndex = stringNumber.lastIndexOf(decimalChar)
    let decimal
    if (decimalIndex > -1) {
      decimal = stringNumber.slice(decimalIndex)
      stringNumber = stringNumber.slice(0, decimalIndex)
    }

    // finally, parse the string. Note, default 'separator' is a comma.
    const parts = parse(stringNumber, optionalSeparator || ',')

    // if there's a decimal value then add it to the parts.
    if (decimal) {// NOTE: we sliced() it off including the decimalChar
      parts.push(decimal)
    }

    // combine all parts for the final string (note, has separators).
    return parts.join('')
  }
}

function parse(string, separator) {

  // find first index to split the string at (where 1st separator goes).
  let i = ((string.length - 1) % 3) + 1

  // above calculation is wrong when num is negative and a certain size.
  if (i === 1 && (string[0] === '-')) {
    i = 4  // example: -123,456,789  start at 4, not 1.
  }

  const strings = [ // holds the string parts
    string.slice(0, i) // grab part before the first separator
  ]

  // split remaining string in groups of 3 where a separator belongs
  for (; i < string.length; i += 3) {
    strings.push(separator, string.substr(i, 3))
  }

  return strings
}


// convenience function for currying style:
//   const format = commaNumber.bindWith(',', '.')
function bindWith(separator, decimalChar) {
  return function(number) {
    return commaNumber(number, separator, decimalChar)
  }
}

module.exports = commaNumber
module.exports.bindWith = bindWith


benchmark/original.js
/**
 * Comma number formatter
 * @param {Number} number Number to format
 * @param {String} [separator=','] Value used to separate numbers
 * @returns {String} Comma formatted number
 */
module.exports = function commaNumber (number, separator) {
  separator = typeof separator === 'undefined' ? ',' : ('' + separator)

  // Convert to number if it's a non-numeric value
  if (typeof number !== 'number') {
    number = Number(number)
  }

  // NaN => 0
  if (isNaN(number)) {
    number = 0
  }

  // Return Infinity immediately
  if (!isFinite(number)) {
    return '' + number
  }

  var stringNumber = ('' + Math.abs(number))
    .split('')
    .reverse()

  var result = []
  for (var i = 0; i < stringNumber.length; i++) {
    if (i && i % 3 === 0) {
      result.push(separator)
    }
    result.push(stringNumber[i])
  }

  // Handle negative numbers
  if (number < 0) {
    result.push('-')
  }

  return result
    .reverse()
    .join('')
}


benchmark/run.js
require('console.table')
var comma = require('../lib/index.js')
  , Benchmark = require('benchmark')

Benchmark.options.initCount  = 10
Benchmark.options.minSamples = 10

// useful when altering things because it runs once per.
// Benchmark.options.initCount = 1
// Benchmark.options.minSamples = 1
// Benchmark.options.minTime = -1
// Benchmark.options.maxTime = -1

inputs = [
  '1',
  '12',
  '123',
  '1234',
  '12345',
  '123456',
  '1234567',
  '12345678',
  '123456789',
  '1234567890',
  '12345678901',
  '123456789012',

  1,
  12,
  123,
  1234,
  12345,
  123456,
  1234567,
  12345678,
  123456789,
  1234567890,
  12345678901,
  123456789012
]

var methods = [
  require('./original'),
  comma
]

var suite = new Benchmark.Suite
  , headers = [' ', 'original', 'alternate']
  , results = []

for (var i = 0; i < inputs.length; i++) {
  var input = inputs[i]
    , name = typeof input === 'string' ? '\'' + input + '\'' : input

  results.push([name])

  suite.add(name, methods[0].bind(null, input))
  suite.add(name, methods[1].bind(null, input))
}

var columns = methods.length + 1
  , row = 0

suite.on('cycle', function(event) {
  var it = event.target
    , which = results[row].length === 1 ? ' original' : 'alternate'

  console.log(which,'completed',it.name)

  results[row].push(comma(it.hz.toFixed(2)) + ' (+-' + it.stats.rme.toFixed(2) + '%)')

  if (results[row].length === columns) {
    return row++
  }
})

suite.on('complete', function() {
  console.log()
  console.table(headers, results)
})

suite.run({
  async: false
})
